ProjFS.Mac/PrjFSLib/prjfs-log/kext-perf-tracing.cpp

```diff
@@ -51,20 +54,34 @@ bool PrjFSLog_FetchAndPrintKextProfilingData(io_connect_t connection)
     }
     else if (ret == kIOReturnSuccess)
     {
-        for (unsigned i = 0; i < Probe_Count; ++i)
+        printf("   Counter                             [ Samples  ][Total time (ns)][Mean (ns)   ][Stddev (ns) ][Min (ns)][Max (ns)  ]\n");
+        printf("----------------------------------------------------------------------------------------------------------------------\n");
+        
+        for (unsigned i = 0; i < PrjFSPerfCounter_Count; ++i)
         {
-            double samples = probes[i].numSamples1;
-            double sum_abs = probes[i].sum;
-            double stddev_abs = samples > 1 ? sqrt((samples * probes[i].sumSquares - sum_abs * sum_abs) / (samples * (samples - 1))) : 0.0;
-
-            double sum_ns = nanosecondsFromAbsoluteTime(sum_abs);
-            double stddev_ns = nanosecondsFromAbsoluteTime(stddev_abs);
-            double mean_ns = samples > 0 ? sum_ns / samples : 0;
-            printf("%2u %40s  %8llu [%8llu] samples, total time: %15.0f ns, mean: %10.2f ns +/- %11.2f",
-                i, PerfCounterNames[i], probes[i].numSamples1, probes[i].numSamples2, sum_ns, mean_ns, stddev_ns);
+            double samples = probes[i].numSamples;
+            printf(
+                "%2u %-35s [%10llu]",
+                i,
+                PerfCounterNames[i],
+                probes[i].numSamples);
+            
             if (probes[i].min != UINT64_MAX)
             {
-                printf(", min: %7llu ns, max: %10llu ns\n",  nanosecondsFromAbsoluteTime(probes[i].min), nanosecondsFromAbsoluteTime(probes[i].max));
+                double sum_abs = probes[i].sum;
+                double stddev_abs = samples > 1 ? sqrt((samples * probes[i].sumSquares - sum_abs * sum_abs) / (samples * (samples - 1))) : 0.0;
```

Comment:
Is this standard deviation, the absolute value of the standard deviation, or absolute deviation? The name is unclear. Also, I'm not confident this formula is correct for any option. Can you double-check?

Reply:
I'll add some comments in here to clarify that, because I had the same questions as I was touching this code. I'm also not entirely convinced that the results are accurate, because often the standard deviations are much bigger than I would expect, up to two orders of magnitude bigger than the mean, which seems counter-intuitive to me. I'll do some more digging here, which includes updating my own knowledge of how to compute standard deviation.

Reply:
That makes sense, but it does make me wonder if these std dev numbers are giving us much value. I certainly don't have an intuitive understanding of what a std dev in a non-Guassian distribution is telling me about our perf. I've been looking at mean and max values to figure out where our hot spots are.

Reply:
Personally, I've found the stddev helpful to get a rough handle on how widely the timings differ, but I agree it's imperfect. I've put together a quick implementation of the logarithmic histogram idea on top of your branch here: https://github.com/pmj/VFSForGit/tree/perfTrace

This changes the perf tracing output to look like this (make sure to scroll all the way to the right):

```
   Counter                             [ Samples  ][Total time (ns)][Mean (ns)   ][Stddev (ns) ][Min (ns)][Max (ns)  ][|←1.00ns  |←1.02µs  |←1.05ms  |←1.07s   |←1100s       Histogram ]
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 0 HandleVnodeOperation                [    475788][      681265058][        1431][       29754][     436][  11131811][        ▁█▇▁▁▁▁▁▁▁▁▁▁▁▁▁                                        ]
 1  |--GetPath                         [    475788][      129222198][         271][         363][      75][    111409][      ▂█▆▁▁▁▁▁▁▁▁                                               ]
 2  |--ShouldHandleVnodeOpEvent        [    475788][      268356515][         564][        1108][     122][    341313][      ▁▁█▇▁▁▁▁▁▁▁▁▁                                             ]
 3  |  |--VnodeIsOnAllowedFilesystem   [    475788][       25934951][          54][         102][      12][     40467][   ▁▅█▄▁▁▁▁▁▁▁ ▁                                                ]
 4  |  |--ShouldIgnoreVnodeType        [    475133][       21197990][          44][         184][      10][    101530][   ▁▅█▃▁▁▁▁▁▁▁ ▁▁                                               ]
 5  |  |  |--Ignored                   [       477]
 6  |  |--TryReadVNodeFileFlags        [    474656][       85522665][         180][         505][      62][    210089][     ▁▆█▃▁▁▁▁▁▁▁▁▁                                              ]
 7  |  |  |--NotInAnyRoot              [    417702]
 8  |  |--IsFileSystemCrawler          [      1857][         155871][          83][          49][      26][      1018][    ▁▄█▂▁▁                                                      ]
 9  |     |--Denied                    [       148]
10  |--TryGetVirtualizationRoot        [      1445][        6319634][        4373][        3274][     904][     61744][         ▁▃█▆▂▁▁                                                ]
11  |  |--FindForVnode                 [      1445][        5263768][        3642][        3186][     310][     60437][        ▁▁▅█▄▁▁▁                                                ]
12  |  |  |--inner_loop_iterations     [      6275][        4663021][         743][        1424][      41][     56047][     ▃▃▁▇█▂▂▁▁▁▁                                                ]
13  |  |--TemporaryDirectory           [         0]
14  |  |--NoRootFound                  [         0]
15  |  |--ProviderOffline              [         0]
16  |  |--CompareProviderPid           [      1445][         205378][         142][          51][      51][       627][     ▁▄█▁▁                                                      ]
17  |     |--OriginatedByProvider      [      1332]
18  |--RaisePreDeleteEvent             [        12][        4529236][      377436][      277913][   99874][    962130][                ▂█▇▄                                            ]
19  |--RaiseEnumerateDirectoryEvent    [        14][       47500378][     3392884][     2773469][  648576][  11125968][                   ▂██▄▂                                        ]
20  |--RaiseRecursivelyEnumerateEvent  [         0]
21  |--RaiseHydrateFileEvent           [        87][      100393820][     1153951][      667531][  457562][   5072089][                  ▂█▆▂▁                                         ]
22 HandleFileOpOperation               [     21481][      205675606][        9574][      326219][     166][  46570774][       ▅█▅▂▁▄▄▂▁▁▁▁▁▁▁▁  ▁                                      ]
23  |--ShouldHandleFileOpEvent         [      7685][      117571393][       15298][      531252][      30][  46569535][    ▁▁▂▁▁▁▁▃██▃▁▁▁▁▁     ▁                                      ]
24  |  |--FindVirtualizationRoot       [      6868][      115651409][       16839][      561938][    2135][  46569227][           ▃██▃▁▁▁▁▁     ▁                                      ]
25  |  |  |--FindForVnode              [      6868][      111902769][       16293][      561921][    1693][  46567466][          ▁▃██▃▁▁▁▁▁     ▁                                      ]
26  |  |  |  |--inner_loop_iterations  [     59151][      107283973][        1813][      191433][      44][  46550770][     ▁▁▁█▅▂▂▁▁▁▁▁▁▁▁     ▁                                      ]
27  |  |  |--NoRootFound               [      6644]
28  |  |--CompareProviderPid           [       224][          39032][         174][          51][      94][       502][      ▁█▁                                                       ]
29  |     |--OriginatedByProvider      [       136]
30  |--RaiseRenamedEvent               [        21][       14823153][      705864][      515161][  285544][   2508603][                  █▅▃▁                                          ]
31  |--RaiseHardLinkCreatedEvent       [         0]
32  |--RaiseFileModifiedEvent          [         7][        2009535][      287076][       99318][  176736][    421548][                 ▆█                                             ]
33  |--RaiseFileCreatedEvent           [        60][       55270221][      921170][      871784][  282938][   5883825][                  █▇▄▁▁                                         ]
```

Feel free to import those commits into this PR, otherwise I'll open my own once you've merged this one.

Incidentally, we can change the bucket resolution if preferred - integer logarithms are a little tricky, but for example doubling the resolution is straightforward: squaring the value before logging to "expand" it to 128 bits gives us 128 buckets.

Reply:
I think that's a cool visualization, but I'm still not really sure how I would use it to make a decision about how to optimize the code. Counts, total times, and mean times seem to be the tool for that job. I'm not opposed to another PR adding more of these visualizations, but I'll take it out of scope for this PR.